File: app/src/main/java/com/http_s/rest/SqliteCore.java
package com.http_s.rest;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;

import org.json.JSONArray;
import org.json.JSONObject;

public class SqliteCore {

    private final String DATABASE_NAME = "httprest.db";
    private final int DATABASE_VERSION = 2;
    private final String SETTINGS_TABLE = "settings";
    private final String HEADERS_VARIABLES_TABLE = "headers_variables";

    private final SQLiteDatabase db;

    private final SQLiteStatement insertStmtSettings;
    private final SQLiteStatement deleteStmtSettings;

    private final SQLiteStatement insertStmtHeadersVariables;
    private final SQLiteStatement deleteStmtHeadersVariables;

    public SqliteCore(Context context) {
        OpenHelper openHelper = new OpenHelper(context);
        this.db = openHelper.getWritableDatabase();
        String INSERT_SETTING = "insert into " + SETTINGS_TABLE + "(setting_name,setting_value) values (?,?)";
        this.insertStmtSettings = this.db.compileStatement(INSERT_SETTING);
        String DELETE_SETTING = "delete from " + SETTINGS_TABLE + " where setting_name = ?";
        this.deleteStmtSettings = this.db.compileStatement(DELETE_SETTING);
        String INSERT_HEADER_VARIABLE = "insert into " + HEADERS_VARIABLES_TABLE + "(field_type, field_name, field_value) values (?,?,?)";
        this.insertStmtHeadersVariables = this.db.compileStatement(INSERT_HEADER_VARIABLE);
        String DELETE_HEADER_VARIABLE = "delete from " + HEADERS_VARIABLES_TABLE + " where field_type = ? and field_name = ?";
        this.deleteStmtHeadersVariables = this.db.compileStatement(DELETE_HEADER_VARIABLE);
    }

    public void closeDb() {
        try {
            db.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    //--------------------   SETTINGS   --------------------
    public String getSetting(String inputSetting) {
        String output = "";

        try {
            Cursor cursor = db.query(SETTINGS_TABLE, new String[]{"setting_value"}, "setting_name = ?", new String[]{inputSetting}, null, null, null);
            if (cursor.moveToFirst()) {
                do {
                    output = cursor.getString(0).trim();
                } while (cursor.moveToNext());
            }

            if (cursor.isClosed()) {
                cursor.close();
            }
        } catch(Exception e) {
            e.printStackTrace();
        }

        return output;
    }

    public void createSetting(String settingName, String settingValue) {
        deleteSetting(settingName);
        this.insertStmtSettings.bindString(1, settingName);
        this.insertStmtSettings.bindString(2, settingValue);
        this.insertStmtSettings.executeInsert();
    }

    public void deleteSetting(String settingName) {
        this.deleteStmtSettings.bindString(1, settingName);
        this.deleteStmtSettings.executeInsert();
    }

    //--------------------   HEADERS / VARIABLES   --------------------
    public JSONArray getHeaderVariable(String inputType, String inputName) {
        JSONArray output = new JSONArray();

        try {
            Cursor cursor;
            if (inputName != null) {
                cursor = db.query(HEADERS_VARIABLES_TABLE, new String[]{"field_name, field_value"}, "field_type = ? and field_name = ?", new String[]{inputType, inputName}, null, null, null);
            } else {
                cursor = db.query(HEADERS_VARIABLES_TABLE, new String[]{"field_name, field_value"}, "field_type = ?", new String[]{inputType}, null, null, null);
            }
            if (cursor.moveToFirst()) {
                do {
                    JSONObject record = new JSONObject();
                    record.put("name", cursor.getString(0).trim());
                    record.put("value", cursor.getString(1).trim());
                    output.put(record);
                } while (cursor.moveToNext());
            }

            if (cursor.isClosed()) {
                cursor.close();
            }
        } catch(Exception e) {
            e.printStackTrace();
        }

        return output;
    }

    public void createHeaderVariable(String fieldType, String fieldName, String fieldValue) {
        deleteHeaderVariable(fieldType, fieldName);
        this.insertStmtHeadersVariables.bindString(1, fieldType);
        this.insertStmtHeadersVariables.bindString(2, fieldName);
        this.insertStmtHeadersVariables.bindString(3, fieldValue);
        this.insertStmtHeadersVariables.executeInsert();
    }

    public void deleteHeaderVariable(String fieldType, String fieldName) {
        this.deleteStmtHeadersVariables.bindString(1, fieldType);
        this.deleteStmtHeadersVariables.bindString(2, fieldName);
        this.deleteStmtHeadersVariables.executeInsert();
    }


    private class OpenHelper extends SQLiteOpenHelper {
        OpenHelper(Context context) {
            super(context, DATABASE_NAME, null, DATABASE_VERSION);
        }

        @Override
        public void onOpen(final SQLiteDatabase db) {

        }

        @Override
        public void onCreate(SQLiteDatabase db) {
            //TRY TO CREATE THE DATABASE TABLES, BUT IT WILL ONLY SUCCEED IF THEY DON'T EXIST
            db.execSQL("CREATE TABLE IF NOT EXISTS " + SETTINGS_TABLE + " (setting_name TEXT, setting_value TEXT)");
            db.execSQL("CREATE TABLE IF NOT EXISTS " + HEADERS_VARIABLES_TABLE + " (field_type TEXT, field_name TEXT, field_value TEXT)");
        }

        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
            onCreate(db);
        }
    }
}

File: app/src/main/java/com/http_s/rest/data/SqliteCore.kt
package com.http_s.rest.data

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteStatement
import org.json.JSONArray
import org.json.JSONObject

class SqliteCore(context: Context?) {
    private val databaseName = "httprest.db"
    private val databaseVersion = 2
    private val settingsTable = "settings"
    private val headerVariablesTable = "headers_variables"
    private val db: SQLiteDatabase
    private val insertStmtSettings: SQLiteStatement
    private val deleteStmtSettings: SQLiteStatement
    private val insertStmtHeadersVariables: SQLiteStatement
    private val deleteStmtHeadersVariables: SQLiteStatement

    init {
        val openHelper = OpenHelper(context)
        db = openHelper.writableDatabase
        insertStmtSettings = db.compileStatement("insert into $settingsTable(setting_name,setting_value) values (?,?)")
        deleteStmtSettings = db.compileStatement("delete from $settingsTable where setting_name = ?")
        insertStmtHeadersVariables = db.compileStatement("insert into $headerVariablesTable(field_type, field_name, field_value) values (?,?,?)")
        deleteStmtHeadersVariables = db.compileStatement("delete from $headerVariablesTable where field_type = ? and field_name = ?")
    }

    fun closeDb() {
        db.close()
    }

    fun getSetting(inputSetting: String): String {
        return db.query(
            settingsTable,
            arrayOf("setting_value"),
            "setting_name = ?",
            arrayOf(inputSetting),
            null,
            null,
            null
        ).use { cursor ->
            if (cursor.moveToFirst()) cursor.getString(0).trim() else ""
        }
    }

    fun createSetting(settingName: String?, settingValue: String?) {
        deleteSetting(settingName)
        insertStmtSettings.bindString(1, settingName.orEmpty())
        insertStmtSettings.bindString(2, settingValue.orEmpty())
        insertStmtSettings.executeInsert()
    }

    private fun deleteSetting(settingName: String?) {
        deleteStmtSettings.bindString(1, settingName.orEmpty())
        deleteStmtSettings.executeUpdateDelete()
    }

    fun getHeaderVariable(inputType: String, inputName: String?): JSONArray {
        val output = JSONArray()
        val selection = if (inputName != null) "field_type = ? and field_name = ?" else "field_type = ?"
        val selectionArgs = if (inputName != null) arrayOf(inputType, inputName) else arrayOf(inputType)
        db.query(
            headerVariablesTable,
            arrayOf("field_name", "field_value"),
            selection,
            selectionArgs,
            null,
            null,
            null
        ).use { cursor ->
            while (cursor.moveToNext()) {
                output.put(
                    JSONObject()
                        .put("name", cursor.getString(0).trim())
                        .put("value", cursor.getString(1).trim())
                )
            }
        }
        return output
    }

    fun createHeaderVariable(fieldType: String?, fieldName: String?, fieldValue: String?) {
        deleteHeaderVariable(fieldType, fieldName)
        insertStmtHeadersVariables.bindString(1, fieldType.orEmpty())
        insertStmtHeadersVariables.bindString(2, fieldName.orEmpty())
        insertStmtHeadersVariables.bindString(3, fieldValue.orEmpty())
        insertStmtHeadersVariables.executeInsert()
    }

    fun deleteHeaderVariable(fieldType: String?, fieldName: String?) {
        deleteStmtHeadersVariables.bindString(1, fieldType.orEmpty())
        deleteStmtHeadersVariables.bindString(2, fieldName.orEmpty())
        deleteStmtHeadersVariables.executeUpdateDelete()
    }

    private inner class OpenHelper(context: Context?) : SQLiteOpenHelper(context, databaseName, null, databaseVersion) {
        override fun onCreate(db: SQLiteDatabase) {
            db.execSQL("CREATE TABLE IF NOT EXISTS $settingsTable (setting_name TEXT, setting_value TEXT)")
            db.execSQL("CREATE TABLE IF NOT EXISTS $headerVariablesTable (field_type TEXT, field_name TEXT, field_value TEXT)")
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            onCreate(db)
        }
    }
}
